subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1616 @@
+static int zms_flash_ate_rd(struct zms_fs *fs, uint64_t addr, struct zms_ate *entry)
+{
+	return zms_flash_rd(fs, addr, entry, sizeof(struct zms_ate));
+}
+
+/* zms_flash_block_cmp compares the data in flash at addr to data
+ * in blocks of size ZMS_BLOCK_SIZE aligned to fs->write_block_size
+ * returns 0 if equal, 1 if not equal, errcode if error
+ */
+static int zms_flash_block_cmp(struct zms_fs *fs, uint64_t addr, const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc;
+	size_t bytes_to_cmp, block_size;
+	uint8_t buf[ZMS_BLOCK_SIZE];
+
+	block_size = ZMS_BLOCK_SIZE & ~(fs->flash_parameters->write_block_size - 1U);
+
+	while (len) {
+		bytes_to_cmp = MIN(block_size, len);
+		rc = zms_flash_rd(fs, addr, buf, bytes_to_cmp);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data8, buf, bytes_to_cmp);
+		if (rc) {
+			return 1;
+		}
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+		data8 += bytes_to_cmp;
+	}
+	return 0;
+}
+
+/* zms_flash_cmp_const compares the data in flash at addr to a constant
+ * value. returns 0 if all data in flash is equal to value, 1 if not equal,
+ * errcode if error
+ */
+static int zms_flash_cmp_const(struct zms_fs *fs, uint64_t addr, uint8_t value, size_t len)
+{
+	int rc;
+	size_t bytes_to_cmp, block_size;
+	uint8_t cmp[ZMS_BLOCK_SIZE];
+
+	block_size = ZMS_BLOCK_SIZE & ~(fs->flash_parameters->write_block_size - 1U);
+
+	(void)memset(cmp, value, block_size);
+	while (len) {
+		bytes_to_cmp = MIN(block_size, len);
+		rc = zms_flash_block_cmp(fs, addr, cmp, bytes_to_cmp);
+		if (rc) {
+			return rc;
+		}
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+	}
+	return 0;
+}
+
+/* flash block move: move a block at addr to the current data write location
+ * and updates the data write location.
+ */
+static int zms_flash_block_move(struct zms_fs *fs, uint64_t addr, size_t len)
+{
+	int rc;
+	size_t bytes_to_copy, block_size;
+	uint8_t buf[ZMS_BLOCK_SIZE];
+
+	block_size = ZMS_BLOCK_SIZE & ~(fs->flash_parameters->write_block_size - 1U);
+
+	while (len) {
+		bytes_to_copy = MIN(block_size, len);
+		rc = zms_flash_rd(fs, addr, buf, bytes_to_copy);
+		if (rc) {
+			return rc;
+		}
+		rc = zms_flash_data_wrt(fs, buf, bytes_to_copy);
+		if (rc) {
+			return rc;
+		}
+		len -= bytes_to_copy;
+		addr += bytes_to_copy;
+	}
+	return 0;
+}
+
+/* erase a sector and verify erase was OK.
+ * return 0 if OK, errorcode on error.
+ */
+static int zms_flash_erase_sector(struct zms_fs *fs, uint64_t addr)
+{
+	int rc;
+	off_t offset;
+
+	addr &= ADDR_SECT_MASK;
+
+	offset = fs->offset;
+	offset += fs->sector_size * (addr >> ADDR_SECT_SHIFT);
+
+	LOG_DBG("Erasing flash at %lx, len %d", (long)offset, fs->sector_size);
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	zms_lookup_cache_invalidate(fs, addr >> ADDR_SECT_SHIFT);
+#endif
+	rc = flash_flatten(fs->flash_device, offset, fs->sector_size);
+
+	if (rc) {
+		return rc;
+	}
+
+	if (zms_flash_cmp_const(fs, addr, fs->flash_parameters->erase_value, fs->sector_size)) {
```

Comment:
This could happen as well very often when using the flash_simulator as we can simulate erase failures (without returning errors) or write failures.
I will add a LOG_ERR message